File: scripts/provider-vocab.py
#!/usr/bin/env python3
"""Print check-provider-boundary.sh's provider-name vocabulary, one word per
line: every plugin catalog.toml publishes, plus every executable name that
plugin's own plugin.toml declares.

Decodes TOML rather than scanning it as text (a commented-out entry or a
name embedded in some other field can't be mistaken for live vocabulary),
using only the standard library so this stays a repo script, not a change
to core.

Usage: provider-vocab.py <plugins-root>
"""
import os
import sys
import tomllib


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: provider-vocab.py <plugins-root>", file=sys.stderr)
        return 2
    plugins_root = sys.argv[1]

    with open(os.path.join(plugins_root, "catalog.toml"), "rb") as f:
        catalog = tomllib.load(f)

    # catalog.toml's `plugins` list is the exact, reviewable set of
    # published plugins (see plugins/catalog.toml's own header comment) —
    # a plugin.toml present on disk but not listed here is not yet
    # published and contributes no vocabulary.
    words = set()
    for rel in catalog["plugins"]:
        words.add(os.path.basename(os.path.normpath(rel)))
        manifest_path = os.path.join(plugins_root, rel, "plugin.toml")
        with open(manifest_path, "rb") as f:
            manifest = tomllib.load(f)
        for executable in manifest.get("executables", []):
            name = executable.get("name")
            if name:
                words.add(name)

    for word in sorted(words):
        print(word)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: app/internal/providervocab/providervocab.go
// Package providervocab derives the provider-name vocabulary
// scripts/check-provider-boundary.sh treats as leaked into core: a shipped
// plugin's directory id, plus every executable name its plugin.toml
// declares. It decodes plugin.toml through the same manifest loader plect
// itself uses, rather than scanning it as text, so a commented-out
// executable or a name embedded in some other field can't be mistaken for
// live vocabulary.
package providervocab

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"github.com/kecbigmt/plecture/app/internal/plugins"
)

// Collect returns the sorted, de-duplicated provider vocabulary declared
// under pluginsRoot: each subdirectory with a plugin.toml contributes its
// directory name (the plugin id) and every name its `[[executables]]`
// declares. A subdirectory without a plugin.toml is not a plugin and is
// skipped.
func Collect(pluginsRoot string) ([]string, error) {
	entries, err := os.ReadDir(pluginsRoot)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", pluginsRoot, err)
	}
	seen := map[string]bool{}
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		dir := filepath.Join(pluginsRoot, entry.Name())
		if _, err := os.Stat(filepath.Join(dir, "plugin.toml")); err != nil {
			continue
		}
		manifest, err := plugins.LoadManifest(dir)
		if err != nil {
			return nil, err
		}
		seen[entry.Name()] = true
		for _, ex := range manifest.Executables {
			seen[ex.Name] = true
		}
	}
	words := make([]string, 0, len(seen))
	for w := range seen {
		words = append(words, w)
	}
	sort.Strings(words)
	return words, nil
}
